src/compute/src/sink/persist_sink.rs

```diff
@@ -52,92 +58,300 @@ where
-    // shard.
+
+    let batch_boundaries = mint_batch_boundaries(
+        sink_id.clone(),
+        operator_name.clone(),
+        &desired_collection.inner,
+        as_of.clone(),
+    );
+
+    let (written_batches, write_token) = write_batches(
+        sink_id.clone(),
+        operator_name.clone(),
+        target,
+        &batch_boundaries,
+        &desired_collection.inner,
+        &persist_collection.inner,
+        as_of.clone(),
+        Arc::clone(&persist_clients),
+    );
+
+    let append_token = append_batches(
+        sink_id.clone(),
+        operator_name.clone(),
+        target,
+        &batch_boundaries,
+        &written_batches,
+        as_of,
+        persist_clients,
+    );
+
+    update_shared_frontier(
+        sink_id.clone(),
+        operator_name,
+        &persist_collection.inner,
+        compute_state,
+    );
+
+    let token = Rc::new((write_token, append_token));
+
+    Some(token)
+}
+
+/// Whenever the frontier advances, this mints a new batch boundary (lower and
+/// upper) that writers should use for writing the next set of batches to
+/// persist.
+///
+/// Only one of the workers does this, meaning there will only be one boundary
+/// in the stream, even in case of multiple timely workers. Use `broadcast()`
+/// to, ahem, broadcast, the one boundary to all downstream write
+/// operators/workers.
+fn mint_batch_boundaries<G>(
+    sink_id: GlobalId,
+    operator_name: String,
+    desired_stream: &Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>,
+    as_of: Antichain<Timestamp>,
+) -> Stream<G, (Antichain<Timestamp>, Antichain<Timestamp>)>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let scope = desired_stream.scope();
+
+    // Only one worker is responsible for determining batch boundaries. All
+    // workers must write batches with the same boundaries, to ensure that they
+    // can be combined into one batch that gets appended to Consensus state.
     let hashed_id = sink_id.hashed();
-    let active_write_worker = (hashed_id as usize) % scope.peers() == scope.index();
+    let active_worker = (hashed_id as usize) % scope.peers() == scope.index();
+
+    let operator_name = format!("{}-determine_batch_boundaries", operator_name);
+
+    let batch_boundaries =
+        desired_stream.unary_frontier(Pipeline, &operator_name, move |cap, _info| {
+            // Only attempt to write from this frontier onward, as our data are
+            // not necessarily correct for times not greater or equal to this
+            // frontier.
+            let mut current_frontier = as_of;
+            let mut cap_set = CapabilitySet::from_elem(cap);
+
+            move |input, output: &mut OutputHandle<u64, _, Tee<u64, _>>| {
+                while let Some((_time, _data)) = input.next() {
+                    // Just read away data.
+                    // WIP: Is this idiomatic timely?
+                }
```

Comment:
This works, but under the hood, the operator receives a full copy of the input stream and immediately discards it. I think there might be a way around this by using [shared](https://github.com/TimelyDataflow/timely-dataflow/blob/master/timely/src/dataflow/operators/rc.rs#L24) operator and consuming the stream of `Rc`s in `mint_batch_boundaries` and `write_batches`.

Reply:
The alternative to avoiding this allocation is to have something like https://github.com/TimelyDataflow/timely-dataflow/pull/439, but that hasn't landed yet.